ferry/includes/importer.py

```diff
@@ -0,0 +1,770 @@
+
+    professors_prep["professors_info"] = professors_prep["professors_info"].apply(
+        zip_professors_info
+    )
+
+    # exclude instances with empty/bad professor infos
+    professors_prep = professors_prep[professors_prep["professors_info"].apply(len) > 0]
+
+    # expand courses with multiple professors
+    professors_prep = professors_prep.loc[
+        :, ["season_code", "course_id", "professors_info"]
+    ].explode("professors_info")
+    professors_prep = professors_prep.reset_index(drop=True)
+
+    # expand professor info columns
+    professors_prep[["name", "email", "ocs_id"]] = pd.DataFrame(
+        professors_prep["professors_info"].tolist(), index=professors_prep.index
+    )
+
+    print("Constructing professors table in chronological order")
+    professors = pd.DataFrame(columns=["professor_id", "name", "email", "ocs_id"])
+
+    professors_by_season = professors_prep.groupby("season_code")
+
+    def get_professor_identifiers(professors):
+        # return dictionaries mapping professors to
+        # professor_id primary keys by names, emails, and ocs_ids
+
+        names_ids = (
+            professors.dropna(subset=["name"])
+            .groupby("name")["professor_id"]
+            .apply(list)
+            .to_dict()
+        )
+        emails_ids = (
+            professors.dropna(subset=["email"])
+            .groupby("email")["professor_id"]
+            .apply(list)
+            .to_dict()
+        )
+        ocs_ids = (
+            professors.dropna(subset=["ocs_id"])
+            .groupby("ocs_id")["professor_id"]
+            .apply(list)
+            .to_dict()
+        )
+
+        return names_ids, emails_ids, ocs_ids
+
+    def match_professors(season_professors, professors):
+        # match professors within a season to main professors
+
+        names_ids, emails_ids, ocs_ids = get_professor_identifiers(professors)
+
+        # get ID matches by field
+        season_professors["name_matched_ids"] = season_professors["name"].apply(
+            lambda x: names_ids.get(x, [])
+        )
+        season_professors["email_matched_ids"] = season_professors["email"].apply(
+            lambda x: emails_ids.get(x, [])
+        )
+        season_professors["ocs_matched_ids"] = season_professors["ocs_id"].apply(
+            lambda x: ocs_ids.get(x, [])
+        )
+
+        # aggregate found IDs
+        season_professors["matched_ids_aggregate"] = (
+            season_professors["name_matched_ids"]
+            + season_professors["email_matched_ids"]
+            + season_professors["ocs_matched_ids"]
+        )
+
+        # aggregate ID matches
+        season_professors["matched_ids_aggregate"] = season_professors[
+            "matched_ids_aggregate"
+        ].apply(lambda x: x if len(x) > 0 else [np.nan])
+
+        # use the most-common matched ID
+        professor_ids = season_professors["matched_ids_aggregate"].apply(
+            lambda x: Counter(x).most_common(1)[0][0]
+        )
```

Comment:
is this stuff implementing the thing that deyuan proposed?

can we spit out warnings when we're uncertain about a given professor <-> ID association?

Reply:
Kind of - this just finds professor IDs on name, email, and ocs_id while accepting the one with the most matches. We can add warnings when there's a tie